src/Kestrel.Core/Internal/Http/HttpProtocol.cs

```diff
@@ -615,6 +483,156 @@ public void OnHeader(Span<byte> name, Span<byte> value)
+                {
+                    await FireOnStarting();
+                }
+
+                PauseStreams();
+
+                if (_onCompleted != null)
+                {
+                    await FireOnCompleted();
+                }
+
+                if (badRequestException == null)
+                {
+                    // If _requestAbort is set, the connection has already been closed.
+                    if (_requestAborted == 0)
+                    {
+                        // Call ProduceEnd() before consuming the rest of the request body to prevent
+                        // delaying clients waiting for the chunk terminator:
+                        //
+                        // https://github.com/dotnet/corefx/issues/17330#issuecomment-288248663
+                        //
+                        // This also prevents the 100 Continue response from being sent if the app
+                        // never tried to read the body.
+                        // https://github.com/aspnet/KestrelHttpServer/issues/2102
+                        //
+                        // ProduceEnd() must be called before _application.DisposeContext(), to ensure
+                        // HttpContext.Response.StatusCode is correctly set when
+                        // IHttpContextFactory.Dispose(HttpContext) is called.
+                        await ProduceEnd();
+
+                        // ForZeroContentLength does not complete the reader nor the writer
+                        if (!messageBody.IsEmpty)
+                        {
+                            try
+                            {
+                                // Finish reading the request body in case the app did not.
+                                await messageBody.ConsumeAsync();
+                            }
+                            catch (BadHttpRequestException ex)
+                            {
+                                // Capture BadHttpRequestException for further processing
+                                badRequestException = ex;
+                            }
+                        }
+                    }
+                    else if (!HasResponseStarted)
+                    {
+                        // If the request was aborted and no response was sent, there's no
+                        // meaningful status code to log.
+                        StatusCode = 0;
+                    }
+                }
+
+                if (badRequestException != null)
+                {
+                    // Handle BadHttpRequestException thrown during app execution or remaining message body consumption.
+                    // This has to be caught here so StatusCode is set properly before disposing the HttpContext
+                    // (DisposeContext logs StatusCode).
+                    SetBadRequestState(badRequestException);
+                }
+
+                application.DisposeContext(httpContext, _applicationException);
```

Comment:
This not being in a finally makes me nervous, @halter73 I'm hoping our tests are complete here.

Reply:
Only thing would be if `await ProduceEnd` could throw a `BadRequestException`

Reply:
Seems like it could? What happens if you set a utf8 response header before the response starts then you throw from the application?

Reply:
They are check when they are set and would throw then in the application. When they are written out they use `PipelineExtensions.WriteAsciiNoValidation`

Reply:
Taking of which `EncodeAsciiCharsToBytes` looks like it could go faster :)

Reply:
I don't think ProduceEnd should ever throw.